internal/clone/registry.go

```diff
@@ -0,0 +1,161 @@
+package clone
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"fmt"
+
+	"github.com/Microsoft/hcsshim/internal/regstate"
+	"github.com/Microsoft/hcsshim/internal/uvm"
+)
+
+const (
+	configRoot = "croot"
+	configKey  = "ckey"
+)
+
+type persistedUVMConfig struct {
+	// actual information related to template / clone
+	RawData []byte
+	// metadata field used to determine if this config is already saved.
+	Stored bool
+}
+
+// When encoding interfaces gob requires us to register the struct types that we will be
+// using under those interfaces. This registration needs to happen on both sides i.e the
+// side which encodes the data and the side which decodes the data.
+// Go init function: https://golang.org/doc/effective_go.html#init
+func init() {
+	// Register the pointer to structs because that is what is being stored.
+	gob.Register(&uvm.VSMBShare{})
+	gob.Register(&uvm.SCSIMount{})
+}
+
+func encodeTemplateConfig(utc *uvm.UVMTemplateConfig) ([]byte, error) {
+	var buf bytes.Buffer
+
+	encoder := gob.NewEncoder(&buf)
+	if err := encoder.Encode(utc); err != nil {
+		return nil, fmt.Errorf("error while encoding template config: %s", err)
+	}
+	return buf.Bytes(), nil
+}
+
+func decodeTemplateConfig(encodedBytes []byte) (*uvm.UVMTemplateConfig, error) {
+	var utc uvm.UVMTemplateConfig
+
+	reader := bytes.NewReader(encodedBytes)
+	decoder := gob.NewDecoder(reader)
+	if err := decoder.Decode(&utc); err != nil {
+		return nil, fmt.Errorf("error while decoding template config: %s", err)
+	}
+	return &utc, nil
+}
+
+// loadPersistedConfig loads a persisted config from the registry that matches the given ID
+// If not found returns `regstate.NotFoundError`
+func loadPersistedUVMConfig(ID string) (*persistedUVMConfig, error) {
+	sk, err := regstate.Open(configRoot, false)
+	if err != nil {
+		return nil, err
+	}
+	defer sk.Close()
+
+	puc := &persistedUVMConfig{
+		Stored: true,
+	}
+	if err := sk.Get(ID, configKey, puc); err != nil {
+		return nil, err
+	}
+	return puc, nil
+}
+
+// storePersistedUVMConfig stores or updates the in-memory config to its registry state.
+// If the store fails returns the store error.
+func storePersistedUVMConfig(ID string, puc *persistedUVMConfig) error {
+	sk, err := regstate.Open(configRoot, false)
+	if err != nil {
+		return err
+	}
+	defer sk.Close()
+
+	if puc.Stored {
```

Comment:
Is there any actual case where we would persist the UVM config twice? If not, we can just treat already existing as an error case.

Reply:
Then we could also get rid of `Stored`, I think.